### PR TITLE
Increase Parallelism in CircleCI and use Shared Workspace

### PR DESCRIPTION
**Pull request by ferristseng**
### Description

 * Increase the number of parallel nodes in CircleCI (we have up to 43)
 * Build webpack in separate job and leverage shared workspace

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,41 +1,9 @@
...
-            POSTGRES_PASSWORD: password
-            POSTGRES_DB: caseflow_certification_test
-
-      # This is our homespun VACOLS container. An oracle db with some special sauce.
-      - image: 008577686731.dkr.ecr.us-gov-west-1.amazonaws.com/facols:latest
 
-      # This is the circleci provided Redis container.
-      - image: circleci/redis:4.0.10
-    resource_class: large
+commands:
```
Comment (ferristseng) [inline .circleci/config.yml:3]: Refactoring some common steps into `command` to make this file more DRY.

File: config/environments/test.rb [hunk 2]
```diff
@@ -61,6 +61,11 @@
     end
   end
 
+  # Allows rake scripts to be run without querying VACOLS on startup
+  if ENV['DISABLE_FACTORY_BOT_INITIALIZERS']
```
Comment (ferristseng) [inline config/environments/test.rb:65]: This fixes a bug in the build process where the "ping VACOLS" step fails because the factory bot initializers try to connect to VACOLS before the actual script body runs.

[Example build where the above happens](https://app.circleci.com/pipelines/github/department-of-veterans-affairs/caseflow/35145/workflows/55239d31-08c1-4206-a459-dbf20ac879e4/jobs/129803/parallel-runs/3/steps/3-113)

Error:

```
#!/bin/bash -eo pipefail
~/project/ci-bin/capture-log "bundle exec rake local:vacols:wait_for_connection"
rake aborted!
OCIError: ORA-01033: ORACLE initialization or shutdown in progress
oci8.c:603:in oci8lib_250.so
/home/circleci/project/vendor/bundle/ruby/2.5.0/gems/ruby-oci8-2.2.7/lib/oci8/oci8.rb:147:in `initialize'
...
/home/circleci/project/app/models/vacols/staff.rb:8:in `<class:Staff>'
/home/circleci/project/app/models/vacols/staff.rb:3:in `<top (required)>'
...
/home/circleci/project/vendor/bundle/ruby/2.5.0/gems/factory_bot-4.11.1/lib/factory_bot/find_definitions.rb:20:in `block (2 levels) in find_definitions'
/home/circleci/project/vendor/bundle/ruby/2.5.0/gems/factory_bot-4.11.1/lib/factory_bot/find_definitions.rb:19:in `each'
...
/home/circleci/project/vendor/bundle/ruby/2.5.0/gems/railties-5.2.4.4/lib/rails/initializable.rb:60:in `run_initializers'
...
/home/circleci/project/vendor/bundle/ruby/2.5.0/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
/usr/local/bin/bundle:30:in `block in <main>'
/usr/local/bin/bundle:22:in `<main>'
Tasks: TOP => local:vacols:wait_for_connection => environment
(See full trace by running task with --trace)

Exited with code exit status 1
CircleCI received exit code 1
```

File: .circleci/config.yml [hunk 3]
```diff
@@ -68,26 +39,105 @@ jobs:
...
           command: |
-            ~/project/ci-bin/capture-log "DB=etl bundle exec rake db:create db:schema:load db:migrate"
-            ~/project/ci-bin/capture-log "bundle exec rake db:create db:schema:load db:migrate"
+            ~/project/ci-bin/capture-log "bundle exec rake local:vacols:wait_for_connection"
+          environment:
+            DISABLE_FACTORY_BOT_INITIALIZERS: 1
+
+jobs:
+  # build workspace
+  build_workspace:
```
Comment (ferristseng) [inline .circleci/config.yml:67]: The biggest performance improvement is using a separate step to build webpack in a [shared workspace](https://circleci.com/docs/2.0/configuration-reference/#persist_to_workspace) that the RSpec executors can leverage in a subsequent job. Previously, each RSpec executor would build the JS bundle whenever a feature test was run. My assumption was that multiple parallel webpack builds on the same instance would bog down the CPU, and slow down test execution drastically.

File: .circleci/config.yml [hunk 4]
```diff
@@ -68,26 +39,105 @@ jobs:
...
+    docker:
+      # The first image listed is the image that all steps run inside of.
+      # This can be modified by steps, or can be built as a separate
+      # customized container.
+      - image: cimg/node:12.3.0 # NodeJS Version defined in ./nvmrc
```
Comment (ferristseng) [inline .circleci/config.yml:72]: This leverages a newer "next generation" docker container that should be considerably lighter than previous ones. I tried to use these new docker images (prefixed with `cimg`) where I could

File: .circleci/config.yml [hunk 5]
```diff
@@ -134,68 +184,28 @@ jobs:
 
       # This is the circleci provided Redis container.
       - image: circleci/redis:4.0.10
-    parallelism: 5
+    parallelism: 9
```
Comment (ferristseng) [inline .circleci/config.yml:187]: Increase the number of parallel nodes to run more tests simultaneously

File: .circleci/config.yml [hunk 6]
```diff
@@ -332,7 +320,7 @@ jobs:
       # The first image listed is the image that all steps run inside of.
       # This can be modified by steps, or can be built as a separate
       # customized container.
-      - image: 008577686731.dkr.ecr.us-gov-west-1.amazonaws.com/circleci:2.5.3-node-browsers
+      - image: circleci/ruby:2.5.3-node-browsers # NodeJS Version defined in ./nvmrc
```
Comment (ferristseng) [inline .circleci/config.yml:323]: I switched this to use a container that CircleCI provided rather than a custom one. The custom one that we build includes an Oracle library that isn't needed for this step. We can eventually just move this to use a next generation container [when they support the `-browsers` variant](https://circleci.com/docs/2.0/circleci-images/#next-gen-language-image-variants).

File: .circleci/config.yml (hunk 3 above)
Comment (alisan16) [reply .circleci/config.yml:67]: Were you able to verify this assumption? Do you have screenshots of the before/after run times?

File: Gemfile [hunk 7]
```diff
@@ -96,7 +96,7 @@ group :test, :development, :demo do
   gem "capybara-screenshot"
   gem "danger", "~> 6.0"
   gem "database_cleaner"
-  gem "factory_bot_rails", "~> 4.8"
+  gem "factory_bot_rails", "~> 5.2"
```
Comment (alisan16) [inline Gemfile:99]: Suggestion: Move gem upgrade to its own PR

File: .circleci/config.yml (hunk 3 above)
Comment (ferristseng) [reply .circleci/config.yml:67]: I'm not sure how I'd verify the CPU usage directly, but I did verify that the RSpec nodes were no longer building the webpack bundle by looking in the RSpec logs for random jobs.

As for build times...

Here is a small random sample of past builds from this morning. I was seeing build times in the range of ~35-50 minutes last week.

Screen Shot 2020-09-21 at 12 09 05 PM

Here is after I made the change *only* to increase the number of parallel executors. Build times improved, but not drastically.

Screen Shot 2020-09-21 at 12 08 00 PM

Here is with both changes. The failed builds are still running all the tests (a single flaky test was the cause of both failures). The build times are ~25-30 minutes with these changes:

Screen Shot 2020-09-21 at 12 08 22 PM

File: Gemfile (hunk 7 above)
Comment (ferristseng) [reply Gemfile:99]: The fix I [added here](https://github.com/department-of-veterans-affairs/caseflow/pull/15274#discussion_r492157914) was added in `5.0` release of `factory_bot_rails`: https://github.com/thoughtbot/factory_bot_rails/issues/192

**Review by alisan16 — APPROVED** [review]
